Miscellaneous additions

pxeger opened the pull request:
For a bunch of small, fairly logically separate, changes like this, is it better to make one big PR like this, or split into many smaller PRs? I thought I'd choose this option so it would be easier to merge, but is that a bad idea because it makes it harder to "pick and choose" changes?

pxeger committed:
instantiate monoid for unit

pxeger committed:
Control: cleanup and code deduplication

pxeger committed:
fix List applicative implementation

pxeger committed:
make definition of Monad_plus consistent

pxeger committed:
add list monoid

pxeger committed:
add tests for string monoid

pxeger committed:
fix typo

pxeger committed:
explicitly annotate that we expect List to satisfy Functor etc.

yallop commented:
One big PR is fine.  It's easy enough to cherry-pick the commits if necessary.

yallop commented on lib/control.ml:
```diff
@@ -44,12 +44,13 @@ let sequence {M : Monad} (ms : 'a M.t list) =
                    val return : 'a -> 'a t
                    val bind : 'a t -> ('a -> 'b t) -> 'b t
-                 end) = struct
+                 end): Monad = struct
```
```suggestion
                 end): Monad with type 'a t = 'a M.t = struct
```

dvlasits commented on lib/control.ml:
```diff
@@ -44,12 +44,13 @@ let sequence {M : Monad} (ms : 'a M.t list) =
-(* Create a monad using default functor and idiom implementation *)
+(* Create a monad using default functor and applicative implementation *)
 module Monad(M : sig
                    type 'a t
```
Why has the syntax changed here to module ```Monad(M : sig)```, from ```module type Monad = sig```?

dvlasits replied:
should this not include applicative?

dvlasits commented on lib/control.ml:
```diff
@@ -44,12 +44,13 @@ let sequence {M : Monad} (ms : 'a M.t list) =
     (return [])
 
-(* Create a monad using default functor and idiom implementation *)
+(* Create a monad using default functor and applicative implementation *)
```
Is this comment not still wrong? Shouldn't it be create Functor and Applicative implementation from Monad implementation?

dvlasits replied:
Ok no my bad, it is still creating a Monad, just letting you sidestep the the Functor and Applicative stages

pxeger committed:
improve Monad factory comment

dvlasits commented on lib/control.ml:
```diff
@@ -128,13 +108,12 @@ implicit module Option = struct
 
   (* Applicative *)
-  let pure x = Some x
+  let return x = Some x
```
It's interesting here that we have return for applicative instead of pure. I see above in the definition of Applicative its defined as return? Is it not better if they're named different. Applicative -> pure and Monad -> return.

dvlasits replied:
Maybe it doesn't matter, but it seems strange to overload the "return" name for multiple modules, how would OCaml know if you do ```return some_applicative```, to expect an applicative and not a monad?

pxeger replied:
`return` must be the same as `pure` by definition, so that function doesn't need to be part of `Monad`. It's a historical accident that there are separate `pure` and `return` functions in Haskell - Applicative didn't used to exist, then it was squeezed in between Functor and Monad, so it needed a different name to `return`. But this suggests (and I agree) that `return` is a better basic name for the concept than `pure`.

dvlasits replied:
I was gonna say I thought ```pure``` is a much better name and that ```return``` is incredibly misleading, but beside the point probably :)

pxeger replied:
I was thinking `return` was "nicer", and it does make sense in `do` notation, but upon reflection `pure` is a bit clearer when used on its own. But yes, this is bikeshedding - I'll just keep it as is to avoid churn.

pxeger committed:
add `with type 'a t = 'a M.t` to Monad

pxeger committed:
improve with type annotation on List

pxeger committed:
fix traverse implementation

pxeger committed:
add tests for list instances

pxeger committed:
clarify the meaning of unqualified "t" in Option

dvlasits commented on lib/control.ml:
```diff
@@ -150,29 +130,37 @@ implicit module Option = struct
+    List.concat (List.map (fun f -> List.map (fun x -> f x) xs) fs)
+
+  (* Monad *)
   let bind x f =
     let rec aux acc = function
       | x :: xs -> aux (x @ acc) xs
       | [] -> acc in
     aux [] (List.rev_map f x)
```
```suggestion
  let bind x f = List.concat (List.map f x)
```

pxeger committed:
Update lib/control.ml

Co-authored-by: Daniel Vlasits <dvlasits@gmail.com>

dvlasits commented on lib/control.ml:
```diff
@@ -150,28 +130,32 @@ implicit module Option = struct
 end
 
-implicit module List = struct
+implicit module List : sig
```
Should there be a similar thing for the option type?

pxeger replied:
Yep I've already done that, just not committed yet

pxeger committed:
Update lib/control.ml

Co-authored-by: Daniel Vlasits <dvlasits@gmail.com>

dvlasits commented on lib/control.ml:
```diff
@@ -184,6 +168,6 @@ implicit module List = struct
   let traverse {F : Applicative} f t =
-    let cons x ys = F.apply (F.apply (F.return (fun x xs -> x :: xs)) (f x)) ys in
-    fold cons t (F.return [])
+    let cons x ys = F.apply (F.fmap (fun x xs -> x :: xs)) (f x) ys in
```
```suggestion
    let cons x ys = F.apply (F.fmap (fun x xs -> x :: xs) (f x)) ys in
```

pxeger committed:
use List.fold_left

pxeger committed:
fix type errors

pxeger committed:
add tests for option instances

pxeger committed:
add signature for Option module

pxeger committed:
split Option and List tests

pxeger committed:
add functor etc. instances for functions and pairs

pxeger commented on lib/any.ml:
```diff
@@ -0,0 +1,7 @@
+
+implicit module Any_Int = struct type t = int end
+implicit module Any_String = struct type t = string end
+implicit module Any_Pair { A : Any } {B : Any } = struct type t = A.t * B.t end
```
@dvlasits Generics could help with this?

dvlasits replied:
What's the goal with this?

dvlasits replied:
It looks kind of like what a Generics code may look like but I'm not sure yet.

pxeger replied:
`Any` is a module type for a module which does nothing except contain a type - so that types can be passed around implicitly to other modules. For example this is needed in https://github.com/modular-implicits/imp/blob/e11a836c67c73d9f5fe83c22a683d1aac8a62e64/lib/control.ml#L201-L205

yallop replied:
This may cause issues down the line. Any module that has a `t` member (e.g. anything that matches `Num` or `Show` or `Monoid`) will match `Any`, so it's very likely that there'll be multiple matching instances whenever `Any` is instantiated.

There's been discussion of a language design that avoids the need for `Any` implicit arguments, by allowing modules with type parameters, but it's not implemented at the moment.

pxeger committed:
add (commented-out due to bug) tests for sequence

pxeger committed:
rename Any.t to Any.t_for_any to avoid potential overlapping instances

pxeger committed:
add Any instance for lists

pxeger committed:
remove duplicate declaration of Num

pxeger committed:
remove duplicate Show classes

pxeger committed:
add show instances for more sizes of tuples

pxeger committed:
strengthen typing of Ord.compare

pxeger committed:
remove ambiguous instantiations of Monoid for numeric types

numbers form many monoids - under addition, multiplication, etc.
so it is a bit weird to choose addition arbitrarily as the "canonical"
Monoid instance for them

pxeger committed:
add type signatures for instances of Eq, Num, Enum, etc.

pxeger committed:
add const and identity functors

pxeger committed:
add myself as author

yallop commented on lib/control.ml:
```diff
@@ -1,4 +1,6 @@
+
+(* Functor, Applicative and Monad module types *)
 module type Functor = sig
   type 'a t
```
```suggestion
  type +'a t
```

(and similarly throughout, since `Monad`, `Functor`, `Applicative`, etc. are all covariant)

yallop commented on lib/control.ml:
```diff
@@ -150,40 +139,90 @@ implicit module Option = struct
+  let fmap m (a, b) = (a, m b)
+end
+
+type ('a, 'b) const = Const of 'a
```
I wonder whether this can be defined simply as
```suggestion
type ('a, 'b) const = 'a
```

If it does need to be a proper (generative) type, it'd be better as a record, so that it's easy to destruct as well as construct:

```ocaml
type ('a, 'b) const = { const: 'a }
```

pxeger replied:
Daniel wants me to merge this so I'm going to make this an issue and deal with it later

yallop commented on lib/data.ml:
```diff
@@ -24,18 +9,24 @@ module Eq = struct
+  let translateCompare (n : int) =
+    if n < 0 then LT
+    else if n = 0 then EQ
+    else (* if n > 0 then *) GT
```
From the name I expected this to be a translation function on comparisons (`(t -> t -> int) -> (t -> t -> ordering)`), and I think it'd be a bit more convenient to use that way

pxeger committed:
instantiate Applicative for Const of a Monoid

pxeger committed:
change type of translateCompare

pxeger committed:
mark Functor et al as covariant
